### PR TITLE
New modules for webapp vnet connection

### LINKED ISSUE #556: Add ability for webapp vnet integration
##### SUMMARY
This is a request for the ability to add [webapp vnet integration](https://docs.microsoft.com/en-us/cli/azure/webapp/vnet-integration?view=azure-cli-latest), so that networking configuration can be defined for a web app.

##### ISSUE TYPE
- Feature Idea

##### COMPONENT NAME

azure_rm_webapp_vnetintegration – Manage Web App Virtual Network Integrations.
azure_rm_webapp_vnetintegration_info – Get Web App Virtual Network Integration facts.

##### ADDITIONAL INFORMATION
This would allow someone to associate a virtual network with a web application.

```yaml
- name: "Configure virtual network for web app."
  azure.azcollection.azure_rm_webapp_vnetintegration:
    resource_group: "MyResourceGroup"
    name: "MyWebapp"
    virtual_network_name: "MyVnetName"
    subnet: "MySubnetName"
```

Please offer thoughts/corrections to above naming. I plan to add a PR to support this feature and would like feedback in advance, if possible!

### PR DESCRIPTION
**Pull request by l3ender**
##### SUMMARY
This PR includes new modules to accomplish ability to configure a webapp's vnet connection (integration), so that networking configuration can be defined for a web app.

Resolves #556.

##### ISSUE TYPE
- New Module Pull Request

##### COMPONENT NAME
azure_rm_webappvnetconnection
azure_rm_webappvnetconnection_info

##### ADDITIONAL INFORMATION

```yaml
- name: Configure web app with virtual network
 azure.azcollection.azure_rm_webappvnetconnection:
    name: "MyWebapp"
    resource_group: "MyResourceGroup"
    vnet_name: "MyVnetName"
    subnet: "MySubnetName"

- name: Configure web app with virtual network in different resource group
  azure.azcollection.azure_rm_webappvnetconnection:
    name: "MyWebapp"
    resource_group: "MyResourceGroup"
    vnet_name: "MyVnetName"
    subnet: "MySubnetName"
    vnet_resource_group: "MyOtherResourceGroup"

- name: Delete web app virtual network
  azure.azcollection.azure_rm_webappvnetconnection:
    name: "MyWebapp"
    resource_group: "MyResourceGroup"
    state: "absent"

- name: View web app vnet connection information
  azure.azcollection.azure_rm_webappvnetconnection_info:
    name: "MyWebapp"
    resource_group: "MyResourceGroup"
```

Output for a connection looks like:
```bash
ok: [localhost] => {
    "msg": {
        "changed": false,
        "connection": {
            "id": "/subscriptions/xxx-xxx/resourceGroups/test-rgproviders/Microsoft.Web/sites/webapp6654533/virtualNetworkConnections/9de64c0a-4bfd-4139-a4bb-2e130df586a1_subnet6654533",
            "name": "9de64c0a-4bfd-4139-a4bb-2e130df586a1_subnet6654533",
            "subnet_name": "subnet6654533",
            "vnet_name": "vnet6654533",
            "vnet_resource_group": "test-rg",
            "vnet_resource_id": "/subscriptions/xxx-xxx/resourceGroups/test-rg/providers/Microsoft.Network/virtualNetworks/vnet6654533/subnets/subnet6654533"
        }
    }
}

```

### CODE REVIEW

**Comment by l3ender** [comment]
@Fred-sun Hello, this is now ready for review. Can you please take a look and advise when you are able?

File: plugins/module_utils/azure_rm_common.py [hunk 1]
```diff
@@ -966,6 +966,14 @@ def get_data_svc_client(self, **kwags):
         config = self.add_user_agent(config)
         return ServiceClient(creds=config.credentials, config=config)
 
+    def get_subnet_detail(self, subnet_id):
+        vnet_detail = subnet_id.split('/Microsoft.Network/virtualNetworks/')[1].split('/subnets/')
+        return dict(
+            resource_group = subnet_id.split('resourceGroups/')[1].split('/')[0],
```
Comment (Fred-sun) [inline plugins/module_utils/azure_rm_common.py]:
```suggestion
            resource_group=subnet_id.split('resourceGroups/')[1].split('/')[0],
```

File: plugins/module_utils/azure_rm_common.py [hunk 2]
```diff
@@ -966,6 +966,14 @@ def get_data_svc_client(self, **kwags):
...
+            vnet_name = vnet_detail[0],
```
Comment (Fred-sun) [inline plugins/module_utils/azure_rm_common.py]:
```suggestion
            vnet_name=vnet_detail[0],
```

File: plugins/module_utils/azure_rm_common.py [hunk 3]
```diff
@@ -966,6 +966,14 @@ def get_data_svc_client(self, **kwags):
...
+            subnet_name = vnet_detail[1],
```
Comment (Fred-sun) [inline plugins/module_utils/azure_rm_common.py]:
```suggestion
            subnet_name=vnet_detail[1],
```

File: plugins/modules/azure_rm_webapp_vnetconnection.py [hunk 4]
```diff
@@ -0,0 +1,254 @@
...
+        - The web app's virtual network connection.
+    returned: always
+    type: complex
+    contains:
+        id:
+            description:
+                - ID of the web app virtual network connection.
+            returned: always
+            type: str
+            sample: /subscriptions/xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx/resourceGroups/myResourceGroup/providers/Microsoft.Web/sites/myWebApp/virtualNetworkConnections/yyyyyyyy-yyyy-yyyy-yyyy-yyyyyyyyyyyy_subnet
```
Comment (Fred-sun) [inline plugins/modules/azure_rm_webapp_vnetconnection.py]: line too long (211 > 160 characters)

File: plugins/modules/azure_rm_webapp_vnetconnection.py [hunk 5]
```diff
@@ -0,0 +1,254 @@
...
+               - Name of the resource group the virtual network is in.
+            returned: always
+            type: str
+            sample: myResourceGroup
+        vnet_resource_id:
+            description:
+               - ID of the virtual network/subnet connected to the web app.
+            returned: always
+            type: str
+            sample: /subscriptions/xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx/resourceGroups/myResourceGroup/providers/Microsoft.Network/virtualNetworks/myVnet/subnets/mySubnet
```
Comment (Fred-sun) [inline plugins/modules/azure_rm_webapp_vnetconnection.py]: line too long (170 > 160 characters)

File: plugins/modules/azure_rm_webapp_vnetconnection.py [hunk 6]
```diff
@@ -0,0 +1,254 @@
...
+'''
+
+try:
+    from azure.mgmt.web.models import SwiftVirtualNetwork
+except Exception:
+    # This is handled in azure_rm_common
+    pass
+
```
Comment (Fred-sun) [inline plugins/modules/azure_rm_webapp_vnetconnection.py]: expected 2 blank lines, found 1
```suggestion


```

File: plugins/modules/azure_rm_webapp_vnetconnection.py [hunk 7]
```diff
@@ -0,0 +1,254 @@
...
+            self.vnet_resource_group = self.vnet_resource_group or self.resource_group
+
+            if not vnet:
+                self.log('Adding vnet connection for webapp {0}'.format(self.name))
+                changed = True
+            else:
+                subnet_detail = self.get_subnet_detail(vnet.vnet_resource_id)
+                if (subnet_detail['resource_group'] != self.vnet_resource_group
+                    or subnet_detail['vnet_name'] != self.vnet_name
+                    or subnet_detail['subnet_name'] != self.subnet):
```
Comment (Fred-sun) [inline plugins/modules/azure_rm_webapp_vnetconnection.py]: visually indented line with same indent as next logical line

File: plugins/modules/azure_rm_webapp_vnetconnection.py [hunk 8]
```diff
@@ -0,0 +1,254 @@
...
+
+    def delete_vnet_connection(self):
+        try:
+            return self.web_client.web_apps.delete_swift_virtual_network(resource_group_name=self.resource_group, name=self.name)
+        except Exception as exc:
+            self.fail("Error deleting webapp vnet connection {0} (rg={1}) - {3}".format(self.name, self.resource_group, str(exc)))
+
+    def create_or_update_vnet_connection(self, vnet):
+        try:
+            return self.web_client.web_apps.create_or_update_swift_virtual_network_connection(resource_group_name=self.resource_group, name=self.name, connection_envelope=vnet)
```
Comment (Fred-sun) [inline plugins/modules/azure_rm_webapp_vnetconnection.py]: line too long (176 > 160 characters)

File: plugins/modules/azure_rm_webapp_vnetconnection.py [hunk 9]
```diff
@@ -0,0 +1,254 @@
...
+        except Exception as exc:
+            self.fail("Error creating/updating webapp vnet connection {0} (vnet={1}, rg={2}) - {3}".format(self.name, self.vnet_name, self.resource_group, str(exc)))
```
Comment (Fred-sun) [inline plugins/modules/azure_rm_webapp_vnetconnection.py]: line too long (165 > 160 characters)

File: plugins/modules/azure_rm_webapp_vnetconnection_info.py [hunk 10]
```diff
@@ -0,0 +1,160 @@
...
+        - The web app's virtual network connection.
+    returned: always
+    type: complex
+    contains:
+        id:
+            description:
+                - ID of the web app virtual network connection.
+            returned: always
+            type: str
+            sample: /subscriptions/xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx/resourceGroups/myResourceGroup/providers/Microsoft.Web/sites/myWebApp/virtualNetworkConnections/yyyyyyyy-yyyy-yyyy-yyyy-yyyyyyyyyyyy_subnet
```
Comment (Fred-sun) [inline plugins/modules/azure_rm_webapp_vnetconnection_info.py]: line too long (211 > 160 characters)

File: plugins/modules/azure_rm_webapp_vnetconnection_info.py [hunk 11]
```diff
@@ -0,0 +1,160 @@
...
+               - Name of the resource group the virtual network is in.
+            returned: always
+            type: str
+            sample: myResourceGroup
+        vnet_resource_id:
+            description:
+               - ID of the virtual network/subnet connected to the web app.
+            returned: always
+            type: str
+            sample: /subscriptions/xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx/resourceGroups/myResourceGroup/providers/Microsoft.Network/virtualNetworks/myVnet/subnets/mySubnet
```
Comment (Fred-sun) [inline plugins/modules/azure_rm_webapp_vnetconnection_info.py]: line too long (170 > 160 characters)

File: plugins/modules/azure_rm_webapp_vnetconnection_info.py [hunk 12]
```diff
@@ -0,0 +1,160 @@
...
+'''
+
```
Comment (Fred-sun) [inline plugins/modules/azure_rm_webapp_vnetconnection_info.py]: expected 2 blank lines, found 1
```suggestion


```

File: plugins/modules/azure_rm_webapp_vnetconnection.py [hunk 13]
```diff
@@ -0,0 +1,254 @@
+#!/usr/bin/python
+#
+# Copyright (c) 2021 Ross Bender (@l3ender)
+#
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
+
+from __future__ import absolute_import, division, print_function
+from ansible_collections.azure.azcollection.plugins.module_utils.azure_rm_common import AzureRMModuleBase
```
Comment (Fred-sun) [inline plugins/modules/azure_rm_webapp_vnetconnection.py]: Import found before documentation variables. All imports must appear below DOCUMENTATION/EXAMPLES/RETURN.

File: plugins/modules/azure_rm_webapp_vnetconnection_info.py [hunk 14]
```diff
@@ -0,0 +1,160 @@
+#!/usr/bin/python
+#
+# Copyright (c) 2021 Ross Bender (@l3ender)
+#
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
+
+from __future__ import absolute_import, division, print_function
+from ansible_collections.azure.azcollection.plugins.module_utils.azure_rm_common import AzureRMModuleBase
```
Comment (Fred-sun) [inline plugins/modules/azure_rm_webapp_vnetconnection_info.py]: Import found before documentation variables. All imports must appear below DOCUMENTATION/EXAMPLES/RETURN.

File: plugins/modules/azure_rm_webapp_vnetconnection.py [hunk 15]
```diff
@@ -0,0 +1,254 @@
...
+__metaclass__ = type
+
+DOCUMENTATION = '''
+---
+module: azure_rm_webapp_vnetconnnection
```
Comment (Fred-sun) [inline plugins/modules/azure_rm_webapp_vnetconnection.py]:
```suggestion
module: azure_rm_webapp_vnetconnection
```

**Comment by Fred-sun** [comment]
@l3ender Please add use cases related to the azure_rm_webapp_vnetconnection_info.py module in tests/integration/targets/azure_rm_webapp_vnetconnection/tasks/main.yml, thank you!

**Comment by Fred-sun** [comment]
@l3ender You are advised to change the module name to --azure_rm_webappvnetconnnection. to ensure that the module name is consistent with other modules. Thank you very much!

**Comment by l3ender** [comment]
@Fred-sun Thank you for the review. I have updated--please let me know if there are other changes!

If can you provide a python linting file which can be used for style/formatting it would be helpful!

File: plugins/modules/azure_rm_webappvnetconnection.py [hunk 16]
```diff
@@ -0,0 +1,259 @@
...
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
+
+from __future__ import absolute_import, division, print_function
+__metaclass__ = type
+
+
+DOCUMENTATION = '''
+---
+module: azure_rm_webappvnetconnection
+version_added: "1.9.0"
```
Comment (Fred-sun) [inline plugins/modules/azure_rm_webappvnetconnection.py]:
```suggestion
version_added: "1.8.0"
```

File: plugins/modules/azure_rm_webappvnetconnection_info.py [hunk 17]
```diff
@@ -0,0 +1,163 @@
...
+
+from __future__ import absolute_import, division, print_function
+__metaclass__ = type
+
+
+DOCUMENTATION = '''
+---
+module: azure_rm_webappvnetconnection_info
+
+version_added: "1.9.0"
```
Comment (Fred-sun) [inline plugins/modules/azure_rm_webappvnetconnection_info.py]:
```suggestion
version_added: "1.8.0"
```

File: plugins/modules/azure_rm_webappvnetconnection_info.py [hunk 18]
```diff
@@ -0,0 +1,163 @@
...
+        self.results = dict(
+            changed=False,
+            connection=dict(),
+        )
+
+        self.name = None
+        self.resource_group = None
+
+        super(AzureRMWebAppVnetConnectionInfo, self).__init__(self.module_arg_spec,
+                                                              supports_tags=False,
```
Comment (Fred-sun) [inline plugins/modules/azure_rm_webappvnetconnection_info.py:115]:
```suggestion
                                                              supports_check_mode=True,
                                                              supports_tags=False,
```

**Review by Fred-sun — COMMENTED** [review]
@l3ender ansible core new request! Thank you very much!

**Comment by xuzhang3** [comment]
@l3ender  can you resolve the file conflicting files

**Comment by l3ender** [comment]
@xuzhang3 Merge conflict resolved...thanks!

**Comment by xuzhang3** [comment]
LGTM

**Comment by l3ender** [comment]
Thanks for merging this (and other PRs), @Fred-sun and @xuzhang3. Is it possible to have a new release including all the latest changes? Thank you!!

**Comment by xuzhang3** [comment]
@l3ender  We are preparing the new release, we still have several tasks to do before the new version.

**Pull request merged**